# STUDENT PERFORMANCE 

*************

# ASSIGNMENT 3

#### I have added a null row and an extra column by the name of "Insta" (Showing that student is on instagram or not) that can be used for data cleaning.

In [1]:
#Import Packages
import numpy as np
import pandas as pd
import pandas_profiling as pp

## Load The data

In [2]:
# Data gets loaded and stored in "data"
data=pd.read_csv(r"Z:\ABES_2YEAR\student_maths_sample.csv")
data

school  sex   age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP    F  18.0       U     GT3       A   4.0   4.0   at_home   teacher   
1       GP    F  17.0       U     GT3       T   1.0   1.0   at_home     other   
2       GP    F  15.0       U     LE3       T   1.0   1.0   at_home     other   
3       GP    F  15.0       U     GT3       T   4.0   2.0    health  services   
4       GP    F  16.0       U     GT3       T   3.0   3.0     other     other   
5       GP    M  16.0       U     LE3       T   4.0   3.0  services     other   
6       GP    M  16.0       U     LE3       T   2.0   2.0     other     other   
7       GP    F  17.0       U     GT3       A   4.0   4.0     other   teacher   
8      NaN  NaN   NaN     NaN     NaN     NaN   NaN   NaN       NaN       NaN   
9       GP    M  15.0       U     LE3       A   3.0   2.0  services     other   
10      GP    M  15.0       U     GT3       T   3.0   4.0     other     other   
11      GP    F  15.0       U     GT3       T   4.0   4.0   teacher    health   
12      GP    F  15.0       U     GT3       T   2.0   1.0  services     other   
13      GP    M  15.0       U     LE3       T   4.0   4.0    health  services   
14      GP    M  15.0       U     GT3       T   4.0   3.0   teacher     other   
15      GP    M  15.0       U     GT3       A   2.0   2.0     other     other   
16      GP    F  16.0       U     GT3       T   4.0   4.0    health     other   
17      GP    F  16.0       U     GT3       T   4.0   4.0  services  services   
18      GP    F  16.0       U     GT3       T   3.0   3.0     other     other   
19      GP    M  17.0       U     GT3       T   3.0   2.0  services  services   
20      GP    M  16.0       U     LE3       T   4.0   3.0    health     other   
21      GP    M  15.0       U     GT3       T   4.0   3.0   teacher     other   
22      GP    M  15.0       U     GT3       T   4.0   4.0    health    health   
23      GP    M  16.0       U     LE3       T   4.0   2.0   teacher     other   
24      GP    M  16.0       U     LE3       T   2.0   2.0     other     other   
25      GP    F  15.0       R     GT3       T   2.0   4.0  services    health   
26      GP    F  16.0       U     GT3       T   2.0   2.0  services  services   
27      GP    M  15.0       U     GT3       T   2.0   2.0     other     other   
28      GP    M  15.0       U     GT3       T   4.0   2.0    health  services   
29      GP    M  16.0       U     LE3       A   3.0   4.0  services     other   
..     ...  ...   ...     ...     ...     ...   ...   ...       ...       ...   
366     MS    M  18.0       R     GT3       T   1.0   3.0   at_home     other   
367     MS    M  18.0       U     LE3       T   4.0   4.0   teacher  services   
368     MS    F  17.0       R     GT3       T   1.0   1.0     other  services   
369     MS    F  18.0       U     GT3       T   2.0   3.0   at_home  services   
370     MS    F  18.0       R     GT3       T   4.0   4.0     other   teacher   
371     MS    F  19.0       U     LE3       T   3.0   2.0  services  services   
372     MS    M  18.0       R     LE3       T   1.0   2.0   at_home  services   
373     MS    F  17.0       U     GT3       T   2.0   2.0     other   at_home   
374     MS    F  17.0       R     GT3       T   1.0   2.0     other     other   
375     MS    F  18.0       R     LE3       T   4.0   4.0     other     other   
376     MS    F  18.0       R     GT3       T   1.0   1.0     other     other   
377     MS    F  20.0       U     GT3       T   4.0   2.0    health     other   
378     MS    F  18.0       R     LE3       T   4.0   4.0   teacher  services   
379     MS    F  18.0       U     GT3       T   3.0   3.0     other     other   
380     MS    F  17.0       R     GT3       T   3.0   1.0   at_home     other   
381     MS    M  18.0       U     GT3       T   4.0   4.0   teacher   teacher   
382     MS    M  18.0       R     GT3       T   2.0   1.0     other     other   
383     MS    M  17.0       U     G

In [3]:
# This shows starting 5 rows of dataset
data.head(5)

school sex   age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F  18.0       U     GT3       A   4.0   4.0  at_home   teacher   
1     GP   F  17.0       U     GT3       T   1.0   1.0  at_home     other   
2     GP   F  15.0       U     LE3       T   1.0   1.0  at_home     other   
3     GP   F  15.0       U     GT3       T   4.0   2.0   health  services   
4     GP   F  16.0       U     GT3       T   3.0   3.0    other     other   

   ... freetime goout  Dalc  Walc  health absences    G1    G2    G3 Insta  
0  ...      3.0   4.0   1.0   1.0     3.0      6.0   5.0   6.0   6.0   yes  
1  ...      3.0   3.0   1.0   1.0     3.0      4.0   5.0   5.0   6.0   yes  
2  ...      3.0   2.0   2.0   3.0     3.0     10.0   7.0   8.0  10.0   yes  
3  ...      2.0   2.0   1.0   1.0     5.0      2.0  15.0  14.0  15.0   yes  
4  ...      3.0   2.0   1.0   2.0     5.0      4.0   6.0  10.0  10.0   yes  

[5 rows x 34 columns]

In [4]:
# This shows last 5 rows of dataset
data.tail(5)

school sex   age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
391     MS   M  20.0       U     LE3       A   2.0   2.0  services  services   
392     MS   M  17.0       U     LE3       T   3.0   1.0  services  services   
393     MS   M  21.0       R     GT3       T   1.0   1.0     other     other   
394     MS   M  18.0       R     LE3       T   3.0   2.0  services     other   
395     MS   M  19.0       U     LE3       T   1.0   1.0     other   at_home   

     ... freetime goout  Dalc  Walc  health absences    G1    G2    G3 Insta  
391  ...      5.0   4.0   4.0   5.0     4.0     11.0   9.0   9.0   9.0    no  
392  ...      4.0   5.0   3.0   4.0     2.0      3.0  14.0  16.0  16.0   yes  
393  ...      5.0   3.0   3.0   3.0     3.0      3.0  10.0   8.0   7.0   yes  
394  ...      4.0   1.0   3.0   4.0     5.0      0.0  11.0  12.0  10.0    no  
395  ...      2.0   3.0   3.0   3.0     5.0      5.0   8.0   9.0   9.0   yes  

[5 rows x 34 columns]

## Describe the data

#### Here, all the attributes and information related to dataset is provided.

In [5]:
# It shows memory usage, datatypes and attributes it consists
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 396 entries, 0 to 395
Data columns (total 34 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null float64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null float64
Fedu          395 non-null float64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null float64
studytime     395 non-null float64
failures      395 non-null float64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null float64
freetime      395 non-null float64
goout         395 non-null float64
Da

In [6]:
# It gives descriptive statistics of dataset given.
data.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

In [7]:
# It gives us description about "school" attribute.
data.school

0       GP
1       GP
2       GP
3       GP
4       GP
5       GP
6       GP
7       GP
8      NaN
9       GP
10      GP
11      GP
12      GP
13      GP
14      GP
15      GP
16      GP
17      GP
18      GP
19      GP
20      GP
21      GP
22      GP
23      GP
24      GP
25      GP
26      GP
27      GP
28      GP
29      GP
      ... 
366     MS
367     MS
368     MS
369     MS
370     MS
371     MS
372     MS
373     MS
374     MS
375     MS
376     MS
377     MS
378     MS
379     MS
380     MS
381     MS
382     MS
383     MS
384     MS
385     MS
386     MS
387     MS
388     MS
389     MS
390     MS
391     MS
392     MS
393     MS
394     MS
395     MS
Name: school, Length: 396, dtype: object

#### Like the above description of "school" attribute, one can also describe other attributes.

In [8]:
# Tells us about the shape of dataset
data.shape

(396, 34)

In [9]:
#Tells us about the name of attributes in dataset
data.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3', 'Insta'],
      dtype='object')

In [10]:
# It gives total size of dataset
data.size

13464

In [11]:
# It shows datatype of each attribute present in dataset
data.dtypes

school         object
sex            object
age           float64
address        object
famsize        object
Pstatus        object
Medu          float64
Fedu          float64
Mjob           object
Fjob           object
reason         object
guardian       object
traveltime    float64
studytime     float64
failures      float64
schoolsup      object
famsup         object
paid           object
activities     object
nursery        object
higher         object
internet       object
romantic       object
famrel        float64
freetime      float64
goout         float64
Dalc          float64
Walc          float64
health        float64
absences      float64
G1            float64
G2            float64
G3            float64
Insta          object
dtype: object

In [12]:
# It gives total number of datatypes used in dataset
data.get_dtype_counts()

float64    16
object     18
dtype: int64

In [13]:
# It gives the total number of non null values in dataset provided.
data.count()

school        395
sex           395
age           395
address       395
famsize       395
Pstatus       395
Medu          395
Fedu          395
Mjob          395
Fjob          395
reason        395
guardian      395
traveltime    395
studytime     395
failures      395
schoolsup     395
famsup        395
paid          395
activities    395
nursery       395
higher        395
internet      395
romantic      395
famrel        395
freetime      395
goout         395
Dalc          395
Walc          395
health        395
absences      395
G1            395
G2            395
G3            395
Insta         395
dtype: int64

In [14]:
# It is used to find the pairwise correlation of all columns in the dataframe
data.corr()

age      Medu      Fedu  traveltime  studytime  failures  \
age         1.000000 -0.163658 -0.163438    0.070641  -0.004140  0.243665   
Medu       -0.163658  1.000000  0.623455   -0.171639   0.064944 -0.236680   
Fedu       -0.163438  0.623455  1.000000   -0.158194  -0.009175 -0.250408   
traveltime  0.070641 -0.171639 -0.158194    1.000000  -0.100909  0.092239   
studytime  -0.004140  0.064944 -0.009175   -0.100909   1.000000 -0.173563   
failures    0.243665 -0.236680 -0.250408    0.092239  -0.173563  1.000000   
famrel      0.053940 -0.003914 -0.001370   -0.016808   0.039731 -0.044337   
freetime    0.016434  0.030891 -0.012846   -0.017025  -0.143198  0.091987   
goout       0.126964  0.064094  0.043105    0.028540  -0.063904  0.124561   
Dalc        0.131125  0.019834  0.002386    0.138325  -0.196019  0.136047   
Walc        0.117276 -0.047123 -0.012631    0.134116  -0.253785  0.141962   
health     -0.062187 -0.046878  0.014742    0.007501  -0.075616  0.065827   
absences    0.175230  0.100285  0.024473   -0.012944  -0.062700  0.063726   
G1         -0.064081  0.205341  0.190270   -0.093040   0.160612 -0.354718   
G2         -0.143474  0.215527  0.164893   -0.153198   0.135880 -0.355896   
G3         -0.161579  0.217147  0.152457   -0.117142   0.097820 -0.360415   

              famrel  freetime     goout      Dalc      Walc    health  \
age         0.053940  0.016434  0.126964  0.131125  0.117276 -0.062187   
Medu       -0.003914  0.030891  0.064094  0.019834 -0.047123 -0.046878   
Fedu       -0.001370 -0.012846  0.043105  0.002386 -0.012631  0.014742   
traveltime -0.016808 -0.017025  0.028540  0.138325  0.134116  0.007501   
studytime   0.039731 -0.143198 -0.063904 -0.196019 -0.253785 -0.075616   
failures   -0.044337  0.091987  0.124561  0.136047  0.141962  0.065827   
famrel      1.000000  0.150701  0.064568 -0.077594 -0.113397  0.094056   
freetime    0.150701  1.000000  0.285019  0.209001  0.147822  0.075733   
goout       0.064568  0.285019  1.000000  0.266994  0.420386 -0.009577   
Dalc       -0.077594  0.209001  0.266994  1.000000  0.647544  0.077180   
Walc       -0.113397  0.147822  0.420386  0.647544  1.000000  0.092476   
health      0.094056  0.075733 -0.009577  0.077180  0.092476  1.000000   
absences   -0.044354 -0.058078  0.044302  0.111908  0.136291 -0.029937   
G1          0.022168  0.012613 -0.149104 -0.094159 -0.126179 -0.073172   
G2         -0.018281 -0.013777 -0.162250 -0.064120 -0.084927 -0.097720   
G3          0.051363  0.011307 -0.132791 -0.054660 -0.051939 -0.061335   

            absences        G1        G2        G3  
age         0.175230 -0.064081 -0.143474 -0.161579  
Medu        0.100285  0.205341  0.215527  0.217147  
Fedu        0.024473  0.190270  0.164893  0.152457  
traveltime -0.012944 -0.093040 -0.153198 -0.117142  
studytime  -0.062700  0.160612  0.135880  0.097820  
failures    0.063726 -0.354718 -0.355896 -0.360415  
famrel     -0.044354  0.022168 -0.018281  0.051363  
freetime   -0.058078  0.012613 -0.013777  0.011307  
goout       0.044302 -0.149104 -0.162250 -0.132791  
Dalc        0.111908 -0.094159 -0.064120 -0.054660  
Walc        0.136291 -0.126179 -0.084927 -0.051939  
health     -0.029937 -0.073172 -0.097720 -0.061335  
absences    1.000000 -0.031003 -0.031777  0.034247  
G1         -0.031003  1.000000  0.852118  0.801468  
G2         -0.031777  0.852118  1.000000  0.904868  
G3          0.034247  0.801468  0.904868  1.000000

In [15]:
data.cov()

age      Medu      Fedu  traveltime  studytime  failures  \
age         1.628285 -0.228619 -0.226949    0.062873  -0.004434  0.231221   
Medu       -0.228619  1.198445  0.742717   -0.131061   0.059667 -0.192681   
Fedu       -0.226949  0.742717  1.184180   -0.120073  -0.008379 -0.202641   
traveltime  0.062873 -0.131061 -0.120073    0.486513  -0.059070  0.047844   
studytime  -0.004434  0.059667 -0.008379   -0.059070   0.704324 -0.108321   
failures    0.231221 -0.192681 -0.202641    0.047844  -0.108321  0.553017   
famrel      0.061717 -0.003842 -0.001337   -0.010512   0.029898 -0.029564   
freetime    0.020947  0.033779 -0.013963   -0.011861  -0.120041  0.068329   
goout       0.180364  0.078115  0.052220    0.022162  -0.059706  0.103123   
Dalc        0.149039  0.019341  0.002313    0.085941  -0.146533  0.090118   
Walc        0.192733 -0.066440 -0.017702    0.120478  -0.274304  0.135964   
health     -0.110326 -0.071349  0.022303    0.007274  -0.088228  0.068059   
absences    1.789501  0.878622  0.213134   -0.072255  -0.421127  0.379265   
G1         -0.271413  0.746135  0.687245   -0.215402   0.447401 -0.875557   
G2         -0.688653  0.887509  0.674953   -0.401940   0.428947 -0.995528   
G3         -0.944612  1.089096  0.760078   -0.374337   0.376110 -1.227932   

              famrel  freetime     goout      Dalc      Walc    health  \
age         0.061717  0.020947  0.180364  0.149039  0.192733 -0.110326   
Medu       -0.003842  0.033779  0.078115  0.019341 -0.066440 -0.071349   
Fedu       -0.001337 -0.013963  0.052220  0.002313 -0.017702  0.022303   
traveltime -0.010512 -0.011861  0.022162  0.085941  0.120478  0.007274   
studytime   0.029898 -0.120041 -0.059706 -0.146533 -0.274304 -0.088228   
failures   -0.029564  0.068329  0.103123  0.090118  0.135964  0.068059   
famrel      0.803997  0.134974  0.064454 -0.061974 -0.130952  0.117252   
freetime    0.134974  0.997725  0.316944  0.185954  0.190163  0.105173   
goout       0.064454  0.316944  1.239388  0.264763  0.602744 -0.014824   
Dalc       -0.061974  0.185954  0.264763  0.793420  0.742852  0.095579   
Walc       -0.130952  0.190163  0.602744  0.742852  1.658678  0.165585   
health      0.117252  0.105173 -0.014824  0.095579  0.165585  1.932944   
absences   -0.318287 -0.464274  0.394718  0.797758  1.404774 -0.333098   
G1          0.065977  0.041817 -0.550967 -0.278385 -0.539388 -0.337666   
G2         -0.061659 -0.051764 -0.679438 -0.214836 -0.411425 -0.511039   
G3          0.211000  0.051745 -0.677292 -0.223061 -0.306464 -0.390677   

             absences         G1         G2         G3  
age          1.789501  -0.271413  -0.688653  -0.944612  
Medu         0.878622   0.746135   0.887509   1.089096  
Fedu         0.213134   0.687245   0.674953   0.760078  
traveltime  -0.072255  -0.215402  -0.401940  -0.374337  
studytime   -0.421127   0.447401   0.428947   0.376110  
failures     0.379265  -0.875557  -0.995528  -1.227932  
famrel      -0.318287   0.065977  -0.061659   0.211000  
freetime    -0.464274   0.041817  -0.051764   0.051745  
goout        0.394718  -0.550967  -0.679438  -0.677292  
Dalc         0.797758  -0.278385  -0.214836  -0.223061  
Walc         1.404774  -0.539388  -0.411425  -0.306464  
health      -0.333098  -0.337666  -0.511039  -0.390677  
absences    64.049541  -0.823556  -0.956596   1.255703  
G1          -0.823556  11.017053  10.638836  12.187682  
G2          -0.956596  10.638836  14.148917  15.593697  
G3           1.255703  12.187682  15.593697  20.989616

## Remove Insignificant columns

#### Here, I will remove two columns("Dalc" and "Insta")

In [16]:
# It deletes the column from your dataset
# If you want to show changes in your original dataset then you can set "inplace" as "True"
drop_list=["Insta","Dalc"]
data.drop(drop_list, inplace=True, axis=1)
data.head(5)

school sex   age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F  18.0       U     GT3       A   4.0   4.0  at_home   teacher   
1     GP   F  17.0       U     GT3       T   1.0   1.0  at_home     other   
2     GP   F  15.0       U     LE3       T   1.0   1.0  at_home     other   
3     GP   F  15.0       U     GT3       T   4.0   2.0   health  services   
4     GP   F  16.0       U     GT3       T   3.0   3.0    other     other   

   ... romantic famrel  freetime  goout  Walc health absences    G1    G2  \
0  ...       no    4.0       3.0    4.0   1.0    3.0      6.0   5.0   6.0   
1  ...       no    5.0       3.0    3.0   1.0    3.0      4.0   5.0   5.0   
2  ...       no    4.0       3.0    2.0   3.0    3.0     10.0   7.0   8.0   
3  ...      yes    3.0       2.0    2.0   1.0    5.0      2.0  15.0  14.0   
4  ...       no    4.0       3.0    2.0   2.0    5.0      4.0   6.0  10.0   

     G3  
0   6.0  
1   6.0  
2  10.0  
3  15.0  
4  10.0  

[5 rows x 32 columns]

## Check Missing value for each columns

In [17]:
# It returns if any value in given dataset is null or not
data.isnull().values.any()

True

In [18]:
# It returns datset with values "True" if it has null value and "False" if value is not null in given dataset 
a=data.isnull()
a.head(10)

school    sex    age  address  famsize  Pstatus   Medu   Fedu   Mjob  \
0   False  False  False    False    False    False  False  False  False   
1   False  False  False    False    False    False  False  False  False   
2   False  False  False    False    False    False  False  False  False   
3   False  False  False    False    False    False  False  False  False   
4   False  False  False    False    False    False  False  False  False   
5   False  False  False    False    False    False  False  False  False   
6   False  False  False    False    False    False  False  False  False   
7   False  False  False    False    False    False  False  False  False   
8    True   True   True     True     True     True   True   True   True   
9   False  False  False    False    False    False  False  False  False   

    Fjob  ...  romantic  famrel  freetime  goout   Walc  health  absences  \
0  False  ...     False   False     False  False  False   False     False   
1  False  ...     False   False     False  False  False   False     False   
2  False  ...     False   False     False  False  False   False     False   
3  False  ...     False   False     False  False  False   False     False   
4  False  ...     False   False     False  False  False   False     False   
5  False  ...     False   False     False  False  False   False     False   
6  False  ...     False   False     False  False  False   False     False   
7  False  ...     False   False     False  False  False   False     False   
8   True  ...      True    True      True   True   True    True      True   
9  False  ...     False   False     False  False  False   False     False   

      G1     G2     G3  
0  False  False  False  
1  False  False  False  
2  False  False  False  
3  False  False  False  
4  False  False  False  
5  False  False  False  
6  False  False  False  
7  False  False  False  
8   True   True   True  
9  False  False  False  

[10 rows x 32 columns]

In [19]:
# It returns number of null values present in each column
data.isnull().sum()

school        1
sex           1
age           1
address       1
famsize       1
Pstatus       1
Medu          1
Fedu          1
Mjob          1
Fjob          1
reason        1
guardian      1
traveltime    1
studytime     1
failures      1
schoolsup     1
famsup        1
paid          1
activities    1
nursery       1
higher        1
internet      1
romantic      1
famrel        1
freetime      1
goout         1
Walc          1
health        1
absences      1
G1            1
G2            1
G3            1
dtype: int64

In [20]:
# It gives the number of null values in present in dataset
data.isnull().sum().sum()

32

## Remove rows with missing values

In [21]:
# It drops the row having all null values in it
data.dropna(how='all', axis=0, inplace=True)
data.head(10)

school sex   age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0      GP   F  18.0       U     GT3       A   4.0   4.0   at_home   teacher   
1      GP   F  17.0       U     GT3       T   1.0   1.0   at_home     other   
2      GP   F  15.0       U     LE3       T   1.0   1.0   at_home     other   
3      GP   F  15.0       U     GT3       T   4.0   2.0    health  services   
4      GP   F  16.0       U     GT3       T   3.0   3.0     other     other   
5      GP   M  16.0       U     LE3       T   4.0   3.0  services     other   
6      GP   M  16.0       U     LE3       T   2.0   2.0     other     other   
7      GP   F  17.0       U     GT3       A   4.0   4.0     other   teacher   
9      GP   M  15.0       U     LE3       A   3.0   2.0  services     other   
10     GP   M  15.0       U     GT3       T   3.0   4.0     other     other   

    ... romantic famrel  freetime  goout  Walc health absences    G1    G2  \
0   ...       no    4.0       3.0    4.0   1.0    3.0      6.0   5.0   6.0   
1   ...       no    5.0       3.0    3.0   1.0    3.0      4.0   5.0   5.0   
2   ...       no    4.0       3.0    2.0   3.0    3.0     10.0   7.0   8.0   
3   ...      yes    3.0       2.0    2.0   1.0    5.0      2.0  15.0  14.0   
4   ...       no    4.0       3.0    2.0   2.0    5.0      4.0   6.0  10.0   
5   ...       no    5.0       4.0    2.0   2.0    5.0     10.0  15.0  15.0   
6   ...       no    4.0       4.0    4.0   1.0    3.0      0.0  12.0  12.0   
7   ...       no    4.0       1.0    4.0   1.0    1.0      6.0   6.0   5.0   
9   ...       no    4.0       2.0    2.0   1.0    1.0      0.0  16.0  18.0   
10  ...       no    5.0       5.0    1.0   1.0    5.0      0.0  14.0  15.0   

      G3  
0    6.0  
1    6.0  
2   10.0  
3   15.0  
4   10.0  
5   15.0  
6   11.0  
7    6.0  
9   19.0  
10  15.0  

[10 rows x 32 columns]

In [22]:
# It reset the index 
# It was needed because the index was also deleted because of deletion of row having null value
a=data.reset_index(drop=True)
a.head(10)

school sex   age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0     GP   F  18.0       U     GT3       A   4.0   4.0   at_home   teacher   
1     GP   F  17.0       U     GT3       T   1.0   1.0   at_home     other   
2     GP   F  15.0       U     LE3       T   1.0   1.0   at_home     other   
3     GP   F  15.0       U     GT3       T   4.0   2.0    health  services   
4     GP   F  16.0       U     GT3       T   3.0   3.0     other     other   
5     GP   M  16.0       U     LE3       T   4.0   3.0  services     other   
6     GP   M  16.0       U     LE3       T   2.0   2.0     other     other   
7     GP   F  17.0       U     GT3       A   4.0   4.0     other   teacher   
8     GP   M  15.0       U     LE3       A   3.0   2.0  services     other   
9     GP   M  15.0       U     GT3       T   3.0   4.0     other     other   

   ... romantic famrel  freetime  goout  Walc health absences    G1    G2  \
0  ...       no    4.0       3.0    4.0   1.0    3.0      6.0   5.0   6.0   
1  ...       no    5.0       3.0    3.0   1.0    3.0      4.0   5.0   5.0   
2  ...       no    4.0       3.0    2.0   3.0    3.0     10.0   7.0   8.0   
3  ...      yes    3.0       2.0    2.0   1.0    5.0      2.0  15.0  14.0   
4  ...       no    4.0       3.0    2.0   2.0    5.0      4.0   6.0  10.0   
5  ...       no    5.0       4.0    2.0   2.0    5.0     10.0  15.0  15.0   
6  ...       no    4.0       4.0    4.0   1.0    3.0      0.0  12.0  12.0   
7  ...       no    4.0       1.0    4.0   1.0    1.0      6.0   6.0   5.0   
8  ...       no    4.0       2.0    2.0   1.0    1.0      0.0  16.0  18.0   
9  ...       no    5.0       5.0    1.0   1.0    5.0      0.0  14.0  15.0   

     G3  
0   6.0  
1   6.0  
2  10.0  
3  15.0  
4  10.0  
5  15.0  
6  11.0  
7   6.0  
8  19.0  
9  15.0  

[10 rows x 32 columns]

In [23]:
# It is used to check if there is any duplicate value present in dataset
data.duplicated().any()

False

# Formatting your Data 

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 395 entries, 0 to 395
Data columns (total 32 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null float64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null float64
Fedu          395 non-null float64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null float64
studytime     395 non-null float64
failures      395 non-null float64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null float64
freetime      395 non-null float64
goout         395 non-null float64
Wa

#### There is no need of formatting data here 

# SUMMARY

### My dataset has 32 attributes.
#### The attributes are : school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Walc, health, absences, G1, G2, G3

### Dataset info:

#### Number of variables	33
#### Number of observations	395
#### Total Missing (%)	0.0%

### Variables types:

#### Numeric	15
#### Categorical	17
#### Boolean	0